### PR TITLE
Use OpenSSL for cert chain verification in CurlHandler

### PR DESCRIPTION
**Pull request by stephentoub**
Our certificate chain verification includes the ability to check user store certificates in addition to the root store, but it's more expensive than the default verification provided by OpenSSL, which only includes the system store.  This commit modifies our CurlHandler's VerifyCertChain implementation to first use the OpenSSL implementation, and only fall back to ours if it fails, resulting in much faster processing in https cases.

(It'll be easiest to look at the whitespace-free diff view: https://github.com/dotnet/corefx/pull/5924/files?w=1 )

cc: @bartonjs

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs [hunk 1]
```diff
@@ -123,54 +123,69 @@ private static void AddChannelBindingToken(X509Certificate2 certificate, IntPtr
...
                 using (SafeX509StoreCtxHandle storeCtx = new SafeX509StoreCtxHandle(storeCtxPtr, ownsHandle: false))
-                using (X509Chain chain = new X509Chain())
                 {
-                    chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
-                    chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
-
-                    IntPtr leafCertPtr = Interop.Crypto.X509StoreCtxGetTargetCert(storeCtx);
-
-                    if (IntPtr.Zero == leafCertPtr)
+                    // First use the default verification provided directly by OpenSSL.
```
Comment (bartonjs) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs]: There are two assumptions here that we might want to codify (in comments):

1) That we don't have a notion of the Windows Disallowed store. (So users can't explicitly un-trust things that the system trusts).
2) That we're not doing CRL checking

If we had a Disallowed store then we'd need to ensure it was empty before doing this (or check after to make sure no Disallowed cert was involved).

If CRL-checking is enabled we'd potentially need to download the CRLs, which is part of our chainwalk logic.

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs]: I'll add a comment.

**Comment by bartonjs** [comment]
LGTM given the promise to codify the assumptions.

Though, we'll probably want to do a similar thing for SslStream.  Since it has the same state of "in theory, the chain is already attached, we just need to see if we trust it".

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test please
Test Innerloop OSX Release Build and Test please

**Comment by bartonjs** [comment]
Test  Innerloop Ubuntu Debug Build and Test please
Test Innerloop Ubuntu Release Build and Test please

**Comment by stephentoub** [comment]
@bartonjs, FYI, there are a few issues that will cause Ubuntu builds to fail right now.  They're being worked on.

**Comment by bartonjs** [comment]
@stephentoub Fair enough, I thought I saw a merge go through, and I was hopeful :)

**Comment by stephentoub** [comment]
> I thought I saw a merge go through

You did. :smile: I just merged a few things where I was confident that no issues were lurking behind the failures, e.g. changing the text string used in a NuGet package description.

**Comment by stephentoub** [comment]
Test Innerloop Ubuntu Debug Build and Test please
Test Innerloop Ubuntu Release Build and Test please

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test please

**Pull request merged**